### DISCUSSION TITLE
stm32_fsdev isochronous support #1249

### DISCUSSION THREAD

**Discussion by vorosj** [hathach/tinyusb#1785] [General]
I read in the comments of the above issue, that there is some progress. skuep wrote:"Hi guys, so I got UAC2 speaker and microphone working on an STM32F302"

Not sure, if the improvements are already merged to the current TinyUSB code, or that is a separate fork?
I am about to write isochronous transfer from host to device for an STM32F303. Just sound from a PC to the built-in ADC.
Should I start from the examples like "audio test" or "uac2_headset", or these are still not working on devices with fsdev, and better to start from skuep's working project? (unfortunately I have no F302, so can't test it in the original form)

This solution is working on other uC-s with fsdev like STM32F103 too?

**Comment by hathach** [comment]
already implemented

**Comment by vorosj** [comment]
Tried the uac2_headset example on stm32f103 and stm32f303. Same problem:
Consider the win11 pc plays youtube music. If I connect the board, win11 selects the board (TinyUSB headset) and silence, so far so good. If I select the normal sound device on win11, and switch back, sometimes youtube drops warning "sound rendering problem" or just stops.
Device manager events: device configured (usbaudio2.inf), device not started (usbaudio2)

**Comment by HiFiPhile** [comment]
Please enable logging and upload log as text file.

**Reply by vorosj** [reply]
Hi,
Thank you for helping.

I compiled it for Segger RTT. (clean, then make BOARD=stm32f103_bluepill LOG=2 LOGGER=rtt all)
Start it without USB cable, then connected the USB to PC.
The behaivor is different then compile without logging. Without logging it enumerates, PC recognise it as soundcard, just fails later. Now, with logging, Win11 tells "USB device not recognized".
 
The log:
00> USBD init on controller 0
00> sizeof(usbd_device_t) = 51
00> sizeof(tu_fifo_t) = 12
00> sizeof(tu_edpt_stream_t) = 24
00> AUDIO init
00> Headset running
00> USBD Bus Reset : Full Speed
00> 
00> USBD Setup Received 80 06 00 01 00 00 40 00 
00>   Get Descriptor Device
00>   Queue EP 80 with 18 bytes ...
00> USBD Xfer Complete on EP 80 with 18 bytes
00>   Queue EP 00 with 0 bytes ...
00> USBD Xfer Complete on EP 00 with 0 bytes
00> USBD Bus Reset : Full Speed
00> 
00> USBD Setup Received 00 05 0B 00 00 00 00 00 
00>   Set Address
00> USBD Xfer Complete on EP 80 with 0 bytes
00> 
00> USBD Setup Received 80 06 00 01 00 00 12 00 
00>   Get Descriptor Device
00>   Queue EP 80 with 18 bytes ...
00> USBD Xfer Complete on EP 80 with 18 bytes
00>   Queue EP 00 with 0 bytes ...
00> USBD Xfer Complete on EP 00 with 0 bytes
00> 
00> USBD Setup Received 80 06 00 02 00 00 FF 00 
00>   Get Descriptor Configuration[0]
00>   Queue EP 80 with 64 bytes ...
00> USBD Xfer Complete on EP 80 with 64 bytes
00>   Queue EP 80 with 64 bytes ...
00> USBD Xfer Complete on EP 80 with 64 bytes
00>   Queue EP 80 with 64 bytes ...
00> USBD Xfer Complete on EP 80 with 64 bytes
00>   Queue EP 80 with 63 bytes ...
00> USBD Xfer Complete on EP 80 with 63 bytes
00>   Queue EP 00 with 0 bytes ...
00> USBD Xfer Complete on EP 00 with 0 bytes
00> 
00> USBD Setup Received 80 06 00 02 00 00 41 01 
00>   Get Descriptor Configuration[0]
00>   Queue EP 80 with 64 bytes ...
00> USBD Xfer Complete on EP 80 with 64 bytes
00>   Queue EP 80 with 64 bytes ...
00> USBD Xfer Complete on EP 80 with 64 bytes
00>   Queue EP 80 with 64 bytes ...
00> USBD Xfer Complete on EP 80 with 64 bytes
00>   Queue EP 80 with 64 bytes ...
00> USBD Xfer Complete on EP 80 with 64 bytes
00>   Queue EP 80 with 64 bytes ...
00> USBD Xfer Complete on EP 80 with 64 bytes
00>   Queue EP 80 with 1 bytes ...
00> USBD Xfer Complete on EP 80 with 1 bytes
00>   Queue EP 00 with 0 bytes ...
00> USBD Xfer Complete on EP 00 with 0 bytes
00> 
00> USBD Setup Received 80 06 03 03 09 04 FF 00 
00>   Get Descriptor String[3]
00>   Queue EP 80 with 14 bytes ...
00> USBD Xfer Complete on EP 80 with 14 bytes
00>   Queue EP 00 with 0 bytes ...
00> USBD Xfer Complete on EP 00 with 0 bytes
00> 
00> USBD Setup Received 80 06 00 03 00 00 FF 00 
00>   Get Descriptor String[0]
00>   Queue EP 80 with 4 bytes ...
00> USBD Xfer Complete on EP 80 with 4 bytes
00>   Queue EP 00 with 0 bytes ...
00> USBD Xfer Complete on EP 00 with 0 bytes
00> 
00> USBD Setup Received 80 06 02 03 09 04 FF 00 
00>   Get Descriptor String[2]
00>   Queue EP 80 with 32 bytes ...
00> USBD Xfer Complete on EP 80 with 32 bytes
00>   Queue EP 00 with 0 bytes ...
00> USBD Xfer Complete on EP 00 with 0 bytes
00> 
00> USBD Setup Received 80 06 00 06 00 00 0A 00 
00>   Get Descriptor Device Qualifier
00>   Stall EP0
00> 
00> USBD Setup Received 80 06 00 01 00 00 12 00 
00>   Get Descriptor Device
00>   Queue EP 80 with 18 bytes ...
00> USBD Xfer Complete on EP 80 with 18 bytes
00>   Queue EP 00 with 0 bytes ...
00> USBD Xfer Complete on EP 00 with 0 bytes
00> 
00> USBD Setup Received 80 06 00 02 00 00 09 00 
00>   Get Descriptor Configuration[0]
00>   Queue EP 80 with 9 bytes ...
00> USBD Xfer Complete on EP 80 with 9 bytes
00>   Queue EP 00 with 0 bytes ...
00> USBD Xfer Complete on EP 00 with 0 bytes
00> 
00> USBD Setup Received 80 06 00 02 00 00 41 01 
00>   Get Descriptor Configuration[0]
00>   Queue EP 80 with 64 bytes ...
00> USBD Xfer Complete on EP 80 with 64 bytes
00>   Queue EP 80 with 64 bytes ...
00> USBD Xfer Complete on EP 80 with 64 bytes
00>   Queue EP 80 with 64 bytes ...
00> USBD Xfer Complete on EP 80 with 64 bytes
00>   Queue EP 80 with 64 bytes ...
00> USBD Xfer Complete on EP 80 with 64 bytes
00>   Queue EP 80 with 64 bytes ...
00> USBD Xfer Complete on EP 80 with 64 bytes
00>   Queue EP 80 with 1 bytes ...
00> USBD Xfer Complete on EP 80 with 1 bytes
00>   Queue EP 00 with 0 bytes ...
00> USBD Xfer Complete on EP 00 with 0 bytes
00> 
00> USBD Setup Received 00 09 01 00 00 00 00 00 
00>   Set Configuration
00> dcd_pma_alloc 771: ASSERT FAILED

**Comment by vorosj** [comment]
Same happened with the stm32f303cc:
00> USBD init on controller 0
00> sizeof(usbd_device_t) = 51
00> sizeof(tu_fifo_t) = 12
00> sizeof(tu_edpt_stream_t) = 24
00> AUDIO init
00> Headset running
00> USBD Bus Reset : Full Speed
00> 
00> USBD Setup Received 80 06 00 01 00 00 40 00 
00>   Get Descriptor Device
00>   Queue EP 80 with 18 bytes ...
00> USBD Xfer Complete on EP 80 with 18 bytes
00>   Queue EP 00 with 0 bytes ...
00> USBD Xfer Complete on EP 00 with 0 bytes
00> USBD Bus Reset : Full Speed
00> 
00> USBD Setup Received 00 05 0B 00 00 00 00 00 
00>   Set Address
00> USBD Xfer Complete on EP 80 with 0 bytes
00> 
00> USBD Setup Received 80 06 00 01 00 00 12 00 
00>   Get Descriptor Device
00>   Queue EP 80 with 18 bytes ...
00> USBD Xfer Complete on EP 80 with 18 bytes
00>   Queue EP 00 with 0 bytes ...
00> USBD Xfer Complete on EP 00 with 0 bytes
00> 
00> USBD Setup Received 80 06 00 02 00 00 FF 00 
00>   Get Descriptor Configuration[0]
00>   Queue EP 80 with 64 bytes ...
00> USBD Xfer Complete on EP 80 with 64 bytes
00>   Queue EP 80 with 64 bytes ...
00> USBD Xfer Complete on EP 80 with 64 bytes
00>   Queue EP 80 with 64 bytes ...
00> USBD Xfer Complete on EP 80 with 64 bytes
00>   Queue EP 80 with 63 bytes ...
00> USBD Xfer Complete on EP 80 with 63 bytes
00>   Queue EP 00 with 0 bytes ...
00> USBD Xfer Complete on EP 00 with 0 bytes
00> 
00> USBD Setup Received 80 06 00 02 00 00 41 01 
00>   Get Descriptor Configuration[0]
00>   Queue EP 80 with 64 bytes ...
00> USBD Xfer Complete on EP 80 with 64 bytes
00>   Queue EP 80 with 64 bytes ...
00> USBD Xfer Complete on EP 80 with 64 bytes
00>   Queue EP 80 with 64 bytes ...
00> USBD Xfer Complete on EP 80 with 64 bytes
00>   Queue EP 80 with 64 bytes ...
00> USBD Xfer Complete on EP 80 with 64 bytes
00>   Queue EP 80 with 64 bytes ...
00> USBD Xfer Complete on EP 80 with 64 bytes
00>   Queue EP 80 with 1 bytes ...
00> USBD Xfer Complete on EP 80 with 1 bytes
00>   Queue EP 00 with 0 bytes ...
00> USBD Xfer Complete on EP 00 with 0 bytes
00> 
00> USBD Setup Received 80 06 03 03 09 04 FF 00 
00>   Get Descriptor String[3]
00>   Queue EP 80 with 14 bytes ...
00> USBD Xfer Complete on EP 80 with 14 bytes
00>   Queue EP 00 with 0 bytes ...
00> USBD Xfer Complete on EP 00 with 0 bytes
00> 
00> USBD Setup Received 80 06 00 03 00 00 FF 00 
00>   Get Descriptor String[0]
00>   Queue EP 80 with 4 bytes ...
00> USBD Xfer Complete on EP 80 with 4 bytes
00>   Queue EP 00 with 0 bytes ...
00> USBD Xfer Complete on EP 00 with 0 bytes
00> 
00> USBD Setup Received 80 06 02 03 09 04 FF 00 
00>   Get Descriptor String[2]
00>   Queue EP 80 with 32 bytes ...
00> USBD Xfer Complete on EP 80 with 32 bytes
00>   Queue EP 00 with 0 bytes ...
00> USBD Xfer Complete on EP 00 with 0 bytes
00> 
00> USBD Setup Received 80 06 00 06 00 00 0A 00 
00>   Get Descriptor Device Qualifier
00>   Stall EP0
00> 
00> USBD Setup Received 80 06 00 01 00 00 12 00 
00>   Get Descriptor Device
00>   Queue EP 80 with 18 bytes ...
00> USBD Xfer Complete on EP 80 with 18 bytes
00>   Queue EP 00 with 0 bytes ...
00> USBD Xfer Complete on EP 00 with 0 bytes
00> 
00> USBD Setup Received 80 06 00 02 00 00 09 00 
00>   Get Descriptor Configuration[0]
00>   Queue EP 80 with 9 bytes ...
00> USBD Xfer Complete on EP 80 with 9 bytes
00>   Queue EP 00 with 0 bytes ...
00> USBD Xfer Complete on EP 00 with 0 bytes
00> 
00> USBD Setup Received 80 06 00 02 00 00 41 01 
00>   Get Descriptor Configuration[0]
00>   Queue EP 80 with 64 bytes ...
00> USBD Xfer Complete on EP 80 with 64 bytes
00>   Queue EP 80 with 64 bytes ...
00> USBD Xfer Complete on EP 80 with 64 bytes
00>   Queue EP 80 with 64 bytes ...
00> USBD Xfer Complete on EP 80 with 64 bytes
00>   Queue EP 80 with 64 bytes ...
00> USBD Xfer Complete on EP 80 with 64 bytes
00>   Queue EP 80 with 64 bytes ...
00> USBD Xfer Complete on EP 80 with 64 bytes
00>   Queue EP 80 with 1 bytes ...
00> USBD Xfer Complete on EP 80 with 1 bytes
00>   Queue EP 00 with 0 bytes ...
00> USBD Xfer Complete on EP 00 with 0 bytes
00> 
00> USBD Setup Received 00 09 01 00 00 00 00 00 
00>   Set Configuration
00> dcd_pma_alloc 771: ASSERT FAILED

**Comment by vorosj** [comment]
Note, that the skuep AIOC (stm32CubeIDE project using his modified TinyUSB version) is properly working for me, and I succesfully finished a project with modding it. (without understanding, why it's working and the very similar uac2_headset sample is not)

**Comment by HiFiPhile** [comment]
Stm32f303cc only have 512 bytes packet FIFO, thus can't fit the packet buffer requirements of uac_headset example.  You can reduce size by looking at how` #defined __RX__` did.
Similar issues is #1935
